src/passes/LocalSubtyping.cpp

```diff
@@ -178,10 +179,37 @@ struct LocalSubtyping : public WalkerPass<PostWalker<LocalSubtyping>> {
         get->type = func->getLocalType(get->index);
       }
       for (auto* set : FindAll<LocalSet>(func->body).list) {
+        auto newType = func->getLocalType(set->index);
         if (set->isTee()) {
-          set->type = func->getLocalType(set->index);
+          set->type = newType;
           set->finalize();
         }
+
+        // If this set was not processed earlier - that is, if it is in
+        // unreachable code - then it may have an incompatible type. That is,
+        // If we saw a reachable set that writes type A, and this set writes
+        // type B, we may have specialized the local type to A, but the value
+        // of type B in this unreachable set is no longer valid to write to
+        // that local. In such a case we must do additional work.
+        if (!Type::isSubType(set->value->type, newType)) {
+          // The type is incompatible. To fix this, replace
+          //
+          //  (set (bad-value))
+          //
+          // with
+          //
+          //  (set (block
+          //   (drop (bad-value))
+          //   (get)
```

Comment:
Why even keep the `set` or even the operand? If this is unreachable code we can just replace the whole expression with `unreachable`. Is it to avoid having to refinalize? If so, how about we replace `(set (bad-value))` with `(nop)` and `(tee (bad-value)` with `(get)`?

Both in my suggestion and in the current code, can the `(get)` in the new pattern cause problems with non-nullable locals enabled?

Reply:
We do still need to keep the value around, as it might have noticeable effects. In particular, a block is reachable if it has any branch to it, and that includes a branch in unreachable code (validation doesn't require first identifying unreachable code and then ignoring branches in it).

We could replace the set, but that would take a little more work (to get pointers to the sets), that I'm not sure is worth it.

It would be simpler to run DCE before this pass, or require DCE to have been run. But we don't have a current mechanism for such things, and it would make this pass less focused to add an internal DCE run before we do anything.

Good point about non-nullable locals and the get... depending on what the final validation rules look like there, this might be an issue. I think we can just put an unreachable there. As we already do a refinalize, that would propagate properly. Done.